src/base/SlipStreamSwapRouter.sol

```diff
@@ -0,0 +1,201 @@
+import {PoolAddressPancakeSwapV3} from "@aperture_finance/uni-v3-lib/src/PoolAddressPancakeSwapV3.sol";
+import {TernaryLib} from "@aperture_finance/uni-v3-lib/src/TernaryLib.sol";
+import {OptimalSwap, TickMath, V3PoolCallee} from "../libraries/OptimalSwap.sol";
+import {PCSV3Immutables, UniV3Immutables} from "./Immutables.sol";
+import {Payments} from "./Payments.sol";
+import "./Callback.sol";
+
+/// @title Optimal Swap Router
+/// @author Aperture Finance
+/// @dev This router swaps through an aggregator to get to approximately the optimal ratio to add liquidity in a UniV3-style
+/// pool, then swaps the tokens to the optimal ratio to add liquidity in the same pool.
+abstract contract SlipStreamSwapRouter is Payments, SlipStreamCallback {
+    using SafeTransferLib for address;
+    using TernaryLib for bool;
+    using TickMath for int24;
+
+    /// @dev Literal numbers used in sqrtPriceLimitX96 = zeroForOne ? MIN_SQRT_RATIO + 1 : MAX_SQRT_RATIO - 1
+    /// = (MAX_SQRT_RATIO - 1) ^ ((MIN_SQRT_RATIO + 1 ^ MAX_SQRT_RATIO - 1) * zeroForOne)
+    uint160 internal constant MAX_SQRT_RATIO_LESS_ONE = 1461446703485210103287273052203988822378723970342 - 1;
+    /// @dev MIN_SQRT_RATIO + 1 ^ MAX_SQRT_RATIO - 1
+    uint160 internal constant XOR_SQRT_RATIO =
+        (4295128739 + 1) ^ (1461446703485210103287273052203988822378723970342 - 1);
+
+    /// @notice Deterministically computes the pool address given the pool key
+    /// @param poolKey The pool key
+    /// @return pool The contract address of the pool
+    function computeAddressSorted(SlipStreamPoolAddress.PoolKey memory poolKey) internal view returns (address pool) {
+        pool = SlipStreamPoolAddress.computeAddressSorted(factory, poolKey);
+    }
+
+    /// @dev Make a direct `exactIn` pool swap
+    /// @param poolKey The pool key containing the token addresses and fee tier
+    /// @param pool The address of the pool
+    /// @param amountIn The amount of token to be swapped
+    /// @param zeroForOne The direction of the swap, true for token0 to token1, false for token1 to token0
+    /// @return amountOut The amount of token received after swap
+    function _poolSwap(
+        SlipStreamPoolAddress.PoolKey memory poolKey,
+        address pool,
+        uint256 amountIn,
+        bool zeroForOne
+    ) internal returns (uint256 amountOut) {
+        if (amountIn != 0) {
+            uint256 wordBeforePoolKey;
+            bytes memory data;
+            assembly ("memory-safe") {
+                // Equivalent to `data = abi.encode(poolKey)`
+                data := sub(poolKey, 0x20)
+                wordBeforePoolKey := mload(data)
+                mstore(data, 0x60)
+            }
+            uint160 sqrtPriceLimitX96;
+            // Equivalent to `sqrtPriceLimitX96 = zeroForOne ? MIN_SQRT_RATIO + 1 : MAX_SQRT_RATIO - 1`
+            assembly {
+                sqrtPriceLimitX96 := xor(MAX_SQRT_RATIO_LESS_ONE, mul(XOR_SQRT_RATIO, zeroForOne))
```

Comment:
for my understanding, if they are equivalent, then why use the mul and xor instead of simply 
`zeroForOne ? MIN_SQRT_RATIO + 1 : MAX_SQRT_RATIO - 1`

Reply:
The assembly code consumes less gas. I personally wouldn't go this far when it comes to gas optimization; this is simply inheriting from existing code.